### PR TITLE
[SPARK-48865][SQL][FOLLOWUP] Add failOnError argument to replace TryEval in TryUrlDecode

### PR DESCRIPTION
**Pull request by wForget**
### What changes were proposed in this pull request?

Add `failOnError` argument for `UrlDecode` to replace `TryEval` in `TryUrlDecode`.

### Why are the changes needed?

Address https://github.com/apache/spark/pull/47294#discussion_r1681150787

> I'm not a big fan of TryEval as it catches all the exceptions, including the ones not from UrlDecode, but from its input expressions.
> 
> We should add a boolean flag in UrlDecode to control the null-on-error behavior.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

existing unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by wForget** [comment]
cc @cloud-fan @yaooqinn

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala [hunk 1]
```diff
@@ -86,16 +86,18 @@ case class UrlEncode(child: Expression)
   since = "3.4.0",
   group = "url_funcs")
 // scalastyle:on line.size.limit
-case class UrlDecode(child: Expression)
+case class UrlDecode(child: Expression, failOnError: Boolean = true)
```
[the remark this answers is not in the document]
Comment (wForget) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:89]: I tested it manually and the newly added parameter did not affect alias.

image

File: sql/core/src/test/resources/sql-tests/analyzer-results/url-functions.sql.out [hunk 2]
```diff
@@ -79,28 +79,28 @@ Project [url_encode(cast(null as string)) AS url_encode(NULL)#x]
 -- !query
 select url_decode('https%3A%2F%2Fspark.apache.org')
 -- !query analysis
-Project [url_decode(https%3A%2F%2Fspark.apache.org) AS url_decode(https%3A%2F%2Fspark.apache.org)#x]
+Project [url_decode(https%3A%2F%2Fspark.apache.org, true) AS url_decode(https%3A%2F%2Fspark.apache.org)#x]
```
Comment (wForget) [inline sql/core/src/test/resources/sql-tests/analyzer-results/url-functions.sql.out:82]: This also shows that the alias has not changed.

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala [hunk 3]
```diff
@@ -86,16 +86,18 @@ case class UrlEncode(child: Expression)
...
+  def this(child: Expression) = this(child, true)
+
   override lazy val replacement: Expression =
     StaticInvoke(
       UrlCodec.getClass,
       SQLConf.get.defaultStringType,
       "decode",
-      Seq(child, Literal("UTF-8")),
-      Seq(StringTypeAnyCollation, StringTypeAnyCollation))
+      Seq(child, Literal("UTF-8"), Literal(failOnError)),
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:99]: Not related to this PR, but `Literal("UTF-8")` can be avoided here

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:99]: hmm how?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:99]: move into the decode method

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:99]: oh, adding an overload of `decode` method that uses `utf8` as the default encoding? SGTM

Comment (wForget) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:99]: Thanks @yaooqinn , I will send a new PR to do it.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:99]: This is the only place we call UrlCodec.decode and the `UTF-8` is the only charset we support

**Pull request closed without merging**